subsys/net/ip/l2/ethernet.c

```diff
@@ -360,5 +530,236 @@ static inline int ethernet_enable(struct net_if *iface, bool state)
+{
+	if (ctx->vlan_enabled) {
+		if (ctx->vlan_enabled == NET_VLAN_MAX_COUNT) {
+			/* All network interface are using VLAN, no need
+			 * to check further.
+			 */
+			return true;
+		}
+
+		if (is_vlan_enabled_for_iface(ctx, iface)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+u16_t net_eth_get_vlan_tag(struct net_if *iface)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	int i;
+
+	for (i = 0; i < CONFIG_NET_VLAN_COUNT; i++) {
+		if (ctx->vlan[i].iface == iface) {
+			return ctx->vlan[i].tag;
+		}
+	}
+
+	return NET_VLAN_TAG_UNSPEC;
+}
+
+static struct ethernet_vlan *get_vlan(struct ethernet_context *ctx,
+				      struct net_if *iface,
+				      u16_t vlan_tag)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NET_VLAN_COUNT; i++) {
+		if (ctx->vlan[i].iface == iface &&
+		    ctx->vlan[i].tag == vlan_tag) {
+			return &ctx->vlan[i];
+		}
+	}
+
+	return NULL;
+}
+
+int net_eth_vlan_enable(struct net_if *iface, u16_t tag)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	const struct ethernet_api *eth =
+		net_if_get_device(iface)->driver_api;
+	struct ethernet_vlan *vlan;
+	int i;
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -EPERM;
+	}
+
+	if (tag == NET_VLAN_TAG_UNSPEC) {
+		return -EBADF;
+	}
+
+	vlan = get_vlan(ctx, iface, tag);
+	if (vlan) {
+		return -EALREADY;
+	}
+
+	for (i = 0; i < CONFIG_NET_VLAN_COUNT; i++) {
+		if (ctx->vlan[i].iface != iface) {
+			continue;
+		}
+
+		if (ctx->vlan[i].tag != NET_VLAN_TAG_UNSPEC) {
+			continue;
+		}
+
+		NET_DBG("[%d] Adding vlan tag %d to iface %p", i, tag, iface);
+
+		ctx->vlan[i].tag = tag;
+
+		enable_vlan_iface(ctx, iface);
+
+		if (eth->vlan_setup) {
+			eth->vlan_setup(iface, tag, true);
+		}
+
+		ctx->vlan_enabled++;
+		if (ctx->vlan_enabled > NET_VLAN_MAX_COUNT) {
+			ctx->vlan_enabled = NET_VLAN_MAX_COUNT;
+		}
+
+		return 0;
+	}
+
+	return -ENOSPC;
+}
+
+int net_eth_vlan_disable(struct net_if *iface, u16_t tag)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	const struct ethernet_api *eth =
+		net_if_get_device(iface)->driver_api;
+	struct ethernet_vlan *vlan;
+
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return -EINVAL;
+	}
+
+	if (tag == NET_VLAN_TAG_UNSPEC) {
+		return -EBADF;
+	}
+
+	vlan = get_vlan(ctx, iface, tag);
+	if (!vlan) {
+		return -ESRCH;
+	}
+
+	NET_DBG("Removing vlan tag %d from iface %p", vlan->tag, vlan->iface);
+
+	vlan->tag = NET_VLAN_TAG_UNSPEC;
+
+	disable_vlan_iface(ctx, iface);
+
+	if (eth->vlan_setup) {
+		eth->vlan_setup(iface, tag, false);
+	}
+
+	ctx->vlan_enabled--;
+	if (ctx->vlan_enabled < 0) {
+		ctx->vlan_enabled = 0;
+	}
+
+	return 0;
+}
+#endif
+
 NET_L2_INIT(ETHERNET_L2, ethernet_recv, ethernet_send, ethernet_reserve,
 	    ethernet_enable);
+
+void ethernet_init(struct net_if *iface)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	int i;
+
+	NET_DBG("Initializing Ethernet L2 %p for iface %p", ctx, iface);
```

Comment:
There is a chicken&egg issue here: drivers call ethernet_init(), even though they do may not support vlan.

For initializing vlan part, maybe you want the driver to tell you via get_hw_caps() if it supports vlan: if so you would run the loop below. Cleaner that way.
So prior to that patch you'll need one that add the capability in the enum (no need to update anything on driver side yet, only your mcux patch will expose that vlan capability afterwards).

Reply:
> There is a chicken&egg issue here: drivers call ethernet_init(), even though they do may not support vlan.

I do not see much issue here. If VLAN is not enabled, then ethernet_init() is just nop. I think we can make the context->is_init variable also VLAN specific and modify the ethernet_init() so that it is really only VLAN specific.

Reply:
> For initializing vlan part, maybe you want the driver to tell you via get_hw_caps() if it supports vlan

Sure, that can be done.